src/main/java/com/google/step/YOUR_PROJECT_NAME_HERE/external/W3SchoolClient.java

```diff
@@ -0,0 +1,42 @@
+package com.google.step.YOUR_PROJECT_NAME_HERE.external;
+
+import com.google.step.YOUR_PROJECT_NAME_HERE.data.Card;
+import java.io.IOException;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.select.Elements;
+
+public final class W3SchoolClient {
+  private static final String CARD_TITLE_ID = "h1";
+  private static final String CARD_DESC_ID = "p";
+  private static final String CARD_SNIPPET_ID = "w3-example";
+  private static final String CARD_CODE_ID = "w3-code";
+
+  public Card search(String w3Link) throws IOException {
+
+    Document doc = Jsoup.connect(w3Link).get();
+    Elements titles = doc.getElementsByTag(CARD_TITLE_ID);
+    Elements descriptions = doc.getElementsByTag(CARD_DESC_ID);
+    Elements snippets = doc.getElementsByClass(CARD_SNIPPET_ID);
+
+    String title = "";
+    String description = "";
+    String code = "";
+
+    if (!isNullOrEmpty(titles) && !isNullOrEmpty(descriptions) && !isNullOrEmpty(snippets)) {
```

Comment:
It really feels like you're trying to avoid doing work unless it's going to be needed later on here.

Completely optional, but have you considered just computing one field at a time, and returning `null` early if it turned out to be null/empty/empty string? e.g.

```java
Element titleElement = doc.getElementsByTag(CARD_TITLE_ID).first();
if (titleElement == null || titleElement.text().isEmpty()) {
  return null;
}
String title = titleElement.text();

// and so forth for the rest, and if you didn't return by the end, return a new card
```